R/sampler-adaptive.R

```diff
@@ -0,0 +1,310 @@
+  ##
+  ## Probably we will provide some method later for extracting this
+  ## internal state and dump it out at the end of the sampling (so
+  ## perhaps 'finalise') as I think we've previously collected this up
+  ## at the end of the sampling as the estimated vcv is of interest.
+  internal <- new.env()
+
+  initialise <- function(state, model, rng) {
+    internal$weight <- 0
+    internal$iteration <- 0
+    
+    internal$mean <- state$pars
+    d <- length(internal$mean)
+    internal$autocorrelation <- matrix(0, d, d)
+    internal$vcv <- update_vcv(internal$mean, internal$autocorrelation,
+                               internal$weight)
+    
+    internal$scaling <- initial_scaling
+    internal$scaling_increment <- scaling_increment %||%
+      calc_scaling_increment(length(internal$mean), acceptance_target,
+                             log_scaling_update)
+    internal$n_start <- calc_n_start(acceptance_target)
+    
+    internal$history_pars <- c()
+    internal$included <- c()
+    internal$scaling_history <- internal$scaling
+  }
+
+  step <- function(state, model, rng) {
+    proposal_vcv <- 
+      calc_proposal_vcv(internal$scaling, internal$vcv, internal$weight,
+                        initial_vcv, initial_vcv_weight)
+    
+    pars_next <- rmvnorm(state$pars, proposal_vcv, rng)
+
+    u <- rng$random_real(1)
+    density_next <- model$density(pars_next)
+    
+    accept_prob <- min(1, exp(density_next - state$density))
+
+    accept <- u < accept_prob
+    if (accept) {
+      state$pars <- pars_next
+      state$density <- density_next
+    }
+
+    internal$iteration <- internal$iteration + 1
+    internal$history_pars <- rbind(internal$history_pars, state$pars) 
+    if (internal$iteration > adapt_end) {
+      internal$scaling_history <- c(internal$scaling_history, internal$scaling)
+      return(state)
+    }
+    is_replacement <- 
+      check_replacement(internal$iteration, forget_rate, forget_end)
+    if (is_replacement) {
+      pars_remove <- internal$history_pars[internal$included[1], ]
+    } else {
+      pars_remove <- NULL
+      internal$weight <- internal$weight + 1 
+    }
+    
+    internal$scaling <- 
+      update_scaling(internal$scaling, internal$iteration, accept_prob,
+                     internal$scaling_increment, min_scaling, acceptance_target, 
+                     pre_diminish, internal$n_start, log_scaling_update)
+    internal$scaling_history <- c(internal$scaling_history, internal$scaling)
+    internal$autocorrelation <- update_autocorrelation(
+      state$pars, internal$weight, internal$autocorrelation, pars_remove)
+    internal$mean <- update_mean(state$pars, internal$weight, internal$mean,
+                                 pars_remove)
+    internal$vcv <- update_vcv(internal$mean, internal$autocorrelation,
+                               internal$weight)
+    internal$included <- 
+      update_included(internal$included, internal$iteration, is_replacement)
```

Comment:
I think this would also be easier to understand with the definition here directly

```suggestion
    if (is_replacement) {
      internal$included <- c(internal$included[-1L], internal$iteration)
    } else {
      internal$included <- c(internal$included, internal$iteration)
    }
```

Reply:
and I think you can then fold it in with the earlier `is_replacement` if/else statement, at which point I'll be a bit clearer on what we're doing with this history saver